stellargraph/layer/deep_graph_infomax.py

```diff
@@ -0,0 +1,137 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from . import GCN, GAT, APPNP, PPNP
+from tensorflow.keras.layers import Input, Dense, Lambda, Layer
+import tensorflow as tf
+from tensorflow.keras import backend as K
+from ..core.experimental import experimental
+
+
+__all__ = [
+    "fullbatch_infomax_node_model",
+    "fullbatch_infomax_embedding_model",
+]
+
+_NODE_FEATS = "GCN_INFO_MAX_NODE_FEATURES"
+
+
+class Discriminator(Layer):
+    """
+    This Layer computes the Discriminator function for Deep Graph Infomax (https://arxiv.org/pdf/1809.10341.pdf).
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def build(self, input_shapes):
+
+        self.kernel = self.add_weight(
+            shape=(input_shapes[0][1], input_shapes[0][1]),
+            initializer="glorot_uniform",
+            name="kernel",
+            regularizer=None,
+            constraint=None,
+        )
+        self.built = True
+
+    def call(self, inputs):
+        """
+        Applies the layer to the inputs.
+
+        Args:
+            inputs: a list of tensors with shapes [(N, F), (F,)] containing the node features and summary feature
+                vector.
+        """
+        features, summary = inputs
+
+        score = tf.linalg.matvec(tf.linalg.matmul(features, self.kernel), summary)
```

Comment:
In the paper, equation 10, applies the sigmoid function to the `score` matrix. I don't see this anywhere in the code, here or in the `fullbatch_infomax_node_model` class where the below lines call the discriminator,

```
 scores = discriminator([node_feats, summary])
 scores_corrupted = discriminator([node_feats_corrupted, summary])
```

Reply:
I moved the application of sigmoid into the loss function for numerical stability, using this model now requires using `tf.nn.sigmoid_cross_entropy_with_logits` - but I should document this somewhere!